File: packages/shallot/src/standard/tumble/engine/table.ts
// Open-addressing hash set of packed shape-pair keys, ported from Box3D's table.c.
//
// The set is a pure membership oracle: the engine only ever asks contains / add / remove, and
// nothing iterates it, so slot and probe order are unobservable. The key layout, fmix hash, linear
// probe, growth threshold, and backward-shift deletion are still ported op-for-op — bit-exact by
// construction is the cheapest thing to verify, and it keeps this file a readable mirror of table.c
// for upstream syncs.
//
// The u64 key is stored as two u32 halves (`keyHi`/`keyLo`) instead of a BigUint64Array, whose every
// read mints a fresh BigInt — one per probe iteration. Hashes live in a parallel Uint32Array; hash 0
// is the empty-slot sentinel, matching the C (which also leaves the collision unguarded).

import { roundUpPowerOf2 } from "./bits";

const SHAPE_MASK = (1 << 22) - 1;
const CHILD_MASK = (1 << 20) - 1;

/**
 * The kernel-resident set arrays' backing. When set, `keyHi`/`keyLo`/`hashes` are views over the
 * kernel's broad-phase region (broadcolumns.ts); `growSet` reserves the region + rewrites the views,
 * and `refreshIfStale` re-derives them if a grow/relocation moved the region. Absent (standalone sets,
 * e.g. tests) → the set owns private Uint32Arrays.
 */
export type SetBacking = { growSet(setCap: number): void; refreshIfStale(): void };

export type HashSet = {
    keyHi: Uint32Array;
    keyLo: Uint32Array;
    hashes: Uint32Array;
    capacity: number;
    count: number;
    // Kernel-resident backing (broadcolumns.ts), or null for a standalone private set. Set → the arrays
    // view the kernel region and the region is reserved lazily on first use (sized to initCapacity).
    store: SetBacking | null;
    // The capacity the first resident reservation sizes to; `capacity` starts 0 until then.
    initCapacity: number;
};

/**
 * High 32 bits of the symmetric shape-pair key. The u64 layout is
 * `[63:42] = min(s1,s2)`, `[41:20] = max(s1,s2)`, `[19:0] = child`, so the larger shape index is
 * the only field straddling bit 32.
 *
 * @example
 * const hi = pairKeyHi(3, 7);
 */
export function pairKeyHi(s1: number, s2: number): number {
    const lo = s1 < s2 ? s1 : s2;
    const hi = s1 < s2 ? s2 : s1;
    return (((lo & SHAPE_MASK) << 10) | ((hi & SHAPE_MASK) >>> 12)) >>> 0;
}

/** Low 32 bits of the symmetric shape-pair key: the larger shape index's low 12 bits, then child. */
export function pairKeyLo(s1: number, s2: number, c: number): number {
    const hi = s1 < s2 ? s2 : s1;
    return (((hi & 0xfff) << 20) | (c & CHILD_MASK)) >>> 0;
}

// Box3D's Murmur3 fmix64 constants. Pair keys are built from increasing integers, so a weak hash
// collides badly.
const K1_HI = 0xff51afd7;
const K1_LO = 0xed558ccd;
const K2_HI = 0xc4ceb9fe;
const K2_LO = 0x1a85ec53;

/**
 * Murmur3's fmix64 over a split key, truncated to the low 32 bits (b3KeyHash). `h ^= h >> 33` on a
 * hi/lo pair is structurally free: it is `lo ^= hi >>> 1` with `hi` unchanged. Each `h *= k` is a
 * 64x64 -> low-64 wrapping multiply over 16-bit limbs, unrolled inline so the accumulator stays in
 * locals — the function is pure and allocates nothing.
 *
 * @example
 * const hash = keyHash(pairKeyHi(3, 7), pairKeyLo(3, 7, 0));
 */
export function keyHash(kHi: number, kLo: number): number {
    let hHi = kHi;
    let hLo = (kLo ^ (kHi >>> 1)) >>> 0;

    let a0 = hLo & 0xffff;
    let a1 = hLo >>> 16;
    let p00 = a0 * (K1_LO & 0xffff);
    let p01 = a0 * (K1_LO >>> 16);
    let p10 = a1 * (K1_LO & 0xffff);
    let p11 = a1 * (K1_LO >>> 16);
    let mid = (p00 >>> 16) + (p01 & 0xffff) + (p10 & 0xffff);
    let outLo = (((mid & 0xffff) << 16) | (p00 & 0xffff)) >>> 0;
    let outHi =
        (p11 +
            (p01 >>> 16) +
            (p10 >>> 16) +
            (mid >>> 16) +
            Math.imul(hLo, K1_HI) +
            Math.imul(hHi, K1_LO)) >>>
        0;

    hHi = outHi;
    hLo = (outLo ^ (outHi >>> 1)) >>> 0;

    a0 = hLo & 0xffff;
    a1 = hLo >>> 16;
    p00 = a0 * (K2_LO & 0xffff);
    p01 = a0 * (K2_LO >>> 16);
    p10 = a1 * (K2_LO & 0xffff);
    p11 = a1 * (K2_LO >>> 16);
    mid = (p00 >>> 16) + (p01 & 0xffff) + (p10 & 0xffff);
    outLo = (((mid & 0xffff) << 16) | (p00 & 0xffff)) >>> 0;
    outHi =
        (p11 +
            (p01 >>> 16) +
            (p10 >>> 16) +
            (mid >>> 16) +
            Math.imul(hLo, K2_HI) +
            Math.imul(hHi, K2_LO)) >>>
        0;

    return (outLo ^ (outHi >>> 1)) >>> 0;
}

export function createSet(capacity: number, store: SetBacking | null = null): HashSet {
    const cap = capacity > 16 ? roundUpPowerOf2(capacity) : 16;
    // Resident set: the arrays live in the kernel's broad-phase region, reserved lazily on first use
    // (which sizes it to initCapacity + zeroes the window so a singleton reused across worlds carries no
    // phantom membership). Starts with empty views + capacity 0.
    if (store !== null) {
        return {
            keyHi: new Uint32Array(0),
            keyLo: new Uint32Array(0),
            hashes: new Uint32Array(0),
            capacity: 0,
            count: 0,
            store,
            initCapacity: cap,
        };
    }
    return {
        keyHi: new Uint32Array(cap),
        keyLo: new Uint32Array(cap),
        hashes: new Uint32Array(cap),
        capacity: cap,
        count: 0,
        store: null,
        initCapacity: cap,
    };
}

// Reserve + zero the resident set's arrays on first use. Idempotent no-op once capacity is set — the
// membership state persists across steps, so this never re-clears a live set. Zeroing the window is what
// makes a fresh world safe over the singleton region (a stale non-zero hash would be phantom membership).
export function ensureResident(set: HashSet): void {
    if (set.store === null) return;
    if (set.capacity === 0) {
        set.capacity = set.initCapacity;
        set.store.growSet(set.capacity); // reserves the region + rewrites keyHi/keyLo/hashes views
        set.keyHi.fill(0);
        set.keyLo.fill(0);
        set.hashes.fill(0);
        return;
    }
    // Already reserved — re-derive the views if a grow/relocation moved the region since the last op.
    set.store.refreshIfStale();
}

export function clearSet(set: HashSet): void {
    ensureResident(set);
    set.count = 0;
    set.keyHi.fill(0);
    set.keyLo.fill(0);
    set.hashes.fill(0);
}

function findSlot(set: HashSet, kHi: number, kLo: number, hash: number): number {
    const mask = set.capacity - 1;
    const hashes = set.hashes;
    const keyHi = set.keyHi;
    const keyLo = set.keyLo;

    let index = hash & mask;
    while (hashes[index] !== 0 && (keyHi[index] !== kHi || keyLo[index] !== kLo)) {
        index = (index + 1) & mask;
    }
    return index;
}

function addKeyHaveCapacity(set: HashSet, kHi: number, kLo: number, hash: number): void {
    const index = findSlot(set, kHi, kLo, hash);
    set.keyHi[index] = kHi;
    set.keyLo[index] = kLo;
    set.hashes[index] = hash;
    set.count += 1;
}

function growTable(set: HashSet): void {
    const oldCapacity = set.capacity;

    if (set.store !== null) {
        // Resident: copy the old contents out to JS scratch *before* reserving (the reserve relocates
        // the arrays and rewrites the views), then reserve the larger region, zero the new window, and
        // re-insert. Membership is preserved; slot/probe order is unobservable, so the rehash is free of
        // bit-exact concern (table.c grows and re-probes identically).
        const oldKeyHi = set.keyHi.slice(0, oldCapacity);
        const oldKeyLo = set.keyLo.slice(0, oldCapacity);
        const oldHashes = set.hashes.slice(0, oldCapacity);

        set.count = 0;
        set.capacity = 2 * oldCapacity;
        set.store.growSet(set.capacity); // reserves the region + rewrites keyHi/keyLo/hashes views
        set.keyHi.fill(0);
        set.keyLo.fill(0);
        set.hashes.fill(0);

        for (let i = 0; i < oldCapacity; ++i) {
            if (oldHashes[i] === 0) {
                continue;
            }
            addKeyHaveCapacity(set, oldKeyHi[i], oldKeyLo[i], oldHashes[i]);
        }
        return;
    }

    const oldKeyHi = set.keyHi;
    const oldKeyLo = set.keyLo;
    const oldHashes = set.hashes;

    set.count = 0;
    set.capacity = 2 * oldCapacity;
    set.keyHi = new Uint32Array(set.capacity);
    set.keyLo = new Uint32Array(set.capacity);
    set.hashes = new Uint32Array(set.capacity);

    for (let i = 0; i < oldCapacity; ++i) {
        if (oldHashes[i] === 0) {
            continue;
        }
        addKeyHaveCapacity(set, oldKeyHi[i], oldKeyLo[i], oldHashes[i]);
    }
}

/** Whether the shape pair is in the set (b3ContainsKey). Symmetric in `s1`/`s2`. */
export function containsKey(set: HashSet, s1: number, s2: number, c: number): boolean {
    ensureResident(set);
    const kHi = pairKeyHi(s1, s2);
    const kLo = pairKeyLo(s1, s2, c);
    const index = findSlot(set, kHi, kLo, keyHash(kHi, kLo));
    return set.keyHi[index] === kHi && set.keyLo[index] === kLo;
}

/** Insert a shape pair. Returns true if it was already present (b3AddKey). */
export function addKey(set: HashSet, s1: number, s2: number, c: number): boolean {
    ensureResident(set);
    const kHi = pairKeyHi(s1, s2);
    const kLo = pairKeyLo(s1, s2, c);
    const hash = keyHash(kHi, kLo);

    const index = findSlot(set, kHi, kLo, hash);
    if (set.hashes[index] !== 0) {
        return true;
    }

    // The C grows only after the miss is confirmed, then re-probes the new table — the pre-grow
    // slot index is stale.
    if (2 * set.count >= set.capacity) {
        growTable(set);
    }
    addKeyHaveCapacity(set, kHi, kLo, hash);
    return false;
}

/** Remove a shape pair. Returns true if it was present. Backward-shifts the probe run to stay dense. */
export function removeKey(set: HashSet, s1: number, s2: number, c: number): boolean {
    ensureResident(set);
    const kHi = pairKeyHi(s1, s2);
    const kLo = pairKeyLo(s1, s2, c);

    let i = findSlot(set, kHi, kLo, keyHash(kHi, kLo));
    if (set.hashes[i] === 0) {
        return false;
    }

    set.keyHi[i] = 0;
    set.keyLo[i] = 0;
    set.hashes[i] = 0;
    set.count -= 1;

    const mask = set.capacity - 1;
    let j = i;
    for (;;) {
        j = (j + 1) & mask;
        if (set.hashes[j] === 0) {
            break;
        }

        // k is the ideal slot for the item at j. Keep it if k lies cyclically in (i, j].
        const k = set.hashes[j] & mask;
        if (i <= j) {
            if (i < k && k <= j) {
                continue;
            }
        } else {
            if (i < k || k <= j) {
                continue;
            }
        }

        set.keyHi[i] = set.keyHi[j];
        set.keyLo[i] = set.keyLo[j];
        set.hashes[i] = set.hashes[j];
        set.keyHi[j] = 0;
        set.keyLo[j] = 0;
        set.hashes[j] = 0;
        i = j;
    }

    return true;
}

File: packages/shallot/rust/tumble/src/table.rs
//! Open-addressing pair-set membership — the read half of box3d's `table.c`, ported for the in-kernel
//! broad-phase dedup (3d). Mirrors `src/table.ts`: the symmetric shape-pair key (two u32 halves), the
//! Murmur3 `fmix64` hash truncated to 32 bits, and the linear-probe `contains`. Integer-only, so it
//! replays the TS hash exactly. The kernel only ever *reads* membership (contact create/destroy still
//! write the set from TS via `addKey`/`removeKey`), so only `contains` is ported.

const SHAPE_MASK: u32 = (1 << 22) - 1;
const CHILD_MASK: u32 = (1 << 20) - 1;

const K1_HI: u32 = 0xff51_afd7;
const K1_LO: u32 = 0xed55_8ccd;
const K2_HI: u32 = 0xc4ce_b9fe;
const K2_LO: u32 = 0x1a85_ec53;

/// High 32 bits of the symmetric shape-pair key (`b3ShapePairKey` hi half).
#[inline]
pub fn pair_key_hi(s1: u32, s2: u32) -> u32 {
    let (lo, hi) = if s1 < s2 { (s1, s2) } else { (s2, s1) };
    ((lo & SHAPE_MASK) << 10) | ((hi & SHAPE_MASK) >> 12)
}

/// Low 32 bits of the symmetric shape-pair key: the larger shape index's low 12 bits, then child.
#[inline]
pub fn pair_key_lo(s1: u32, s2: u32, c: u32) -> u32 {
    let hi = if s1 < s2 { s2 } else { s1 };
    ((hi & 0xfff) << 20) | (c & CHILD_MASK)
}

/// Murmur3 `fmix64` over the split key, truncated to the low 32 bits (b3KeyHash). Ported op-for-op from
/// `src/table.ts::keyHash`; `Math.imul` maps to `wrapping_mul` on u32. Each `h *= k` round is a 64x64 →
/// low-64 wrapping multiply over 16-bit limbs (`round`).
pub fn key_hash(k_hi: u32, k_lo: u32) -> u32 {
    let mut h_hi = k_hi;
    let mut h_lo = k_lo ^ (k_hi >> 1);

    // Round 1 (K1).
    let (out_lo, out_hi) = round(h_lo, h_hi, K1_LO, K1_HI);
    h_hi = out_hi;
    h_lo = out_lo ^ (out_hi >> 1);

    // Round 2 (K2).
    let (out_lo, out_hi) = round(h_lo, h_hi, K2_LO, K2_HI);

    out_lo ^ (out_hi >> 1)
}

#[inline]
fn round(h_lo: u32, h_hi: u32, k_lo: u32, k_hi: u32) -> (u32, u32) {
    let a0 = h_lo & 0xffff;
    let a1 = h_lo >> 16;
    let p00 = a0.wrapping_mul(k_lo & 0xffff);
    let p01 = a0.wrapping_mul(k_lo >> 16);
    let p10 = a1.wrapping_mul(k_lo & 0xffff);
    let p11 = a1.wrapping_mul(k_lo >> 16);
    let mid = (p00 >> 16)
        .wrapping_add(p01 & 0xffff)
        .wrapping_add(p10 & 0xffff);
    let out_lo = ((mid & 0xffff) << 16) | (p00 & 0xffff);
    let out_hi = p11
        .wrapping_add(p01 >> 16)
        .wrapping_add(p10 >> 16)
        .wrapping_add(mid >> 16)
        .wrapping_add(h_lo.wrapping_mul(k_hi))
        .wrapping_add(h_hi.wrapping_mul(k_lo));
    (out_lo, out_hi)
}

/// Whether the shape pair `(s1, s2, c)` is in the set (b3ContainsKey). Symmetric in `s1`/`s2`.
/// `cap` is a power of two (probe mask `cap - 1`); `key_hi`/`key_lo`/`hashes` are the set's arrays.
pub fn contains(
    key_hi: &[u32],
    key_lo: &[u32],
    hashes: &[u32],
    cap: usize,
    s1: u32,
    s2: u32,
    c: u32,
) -> bool {
    if cap == 0 {
        return false;
    }
    let k_hi = pair_key_hi(s1, s2);
    let k_lo = pair_key_lo(s1, s2, c);
    let hash = key_hash(k_hi, k_lo);
    let mask = (cap - 1) as u32;
    let mut idx = (hash & mask) as usize;
    while hashes[idx] != 0 && (key_hi[idx] != k_hi || key_lo[idx] != k_lo) {
        idx = ((idx as u32 + 1) & mask) as usize;
    }
    key_hi[idx] == k_hi && key_lo[idx] == k_lo
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn key_hash_matches_ts() {
        // (s1, s2, c, expected_hash) from src/table.ts (TS reference).
        let cases = [
            (3u32, 7u32, 0u32, 3489682763u32),
            (0, 10, 0, 2285512313),
            (2, 5, 0, 1184001114),
            (100, 3, 0, 179916708),
            (0, 1, 0, 1985457684),
        ];
        for (a, b, c, want) in cases {
            let hi = pair_key_hi(a, b);
            let lo = pair_key_lo(a, b, c);
            assert_eq!(key_hash(hi, lo), want, "keyHash({a},{b},{c})");
        }
    }
}
